docs/design/operator.md

```diff
@@ -0,0 +1,200 @@
+acknowledged that this may not be immediately possible due to restrictions in
+the underlying technology but an effort will still be made.
+
+### Deployment
+
+The Gluster operator will be able to deploy a new Gluster cluster using a set of
+configuration parameters. This configuration will allow for extensive
+customization if desired but will assume sane defaults for as many parameters as
+possible to allow for as minimal a configuration as possible.
+
+**Target Configurations:** Hosted, Managed
+
+### Import
+
+The Gluster operator will be able to import an existing Gluster cluster at any
+stage of initialization. The operator's desired configuration for the cluster
+will need to be compatible with the state of the cluster at the time of import.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Cluster Upgrade
+
+The Gluster operator will be able to execute a rolling upgrade of any hosted or
+managed Gluster cluster. It will trigger an upgrade one node at a time, waiting
+for the node to come up and for the cluster to report no pending heals to any
+files before triggering a subsequent node.
+
+**Target Configurations:** Hosted, Managed
+
+### Monitoring
+
+To provide event-driven features, the Gluster operator will monitor various
+aspects of the clusters it will automate.
+
+#### Storage Capacity
+
+The Gluster operator will monitor the storage capacity of the overall cluster.
+Storage capacity is defined here as the amount of space that can be committed to
+new Gluster volumes and not how much free space is available. It will trigger
+events when storage capacity reaches "Warning" levels and "Critical" levels.
+These levels will be configurable, and should be set to a value that will allow
+for corrective action to take place before the storage reaches maximum capacity.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Volume Density
+
+The Gluster operator will monitor the total volume density of any cluster it is
+automating. It will trigger events when the number of volumes in the cluster
+reaches a pre-defined level that should prompt corrective action. This value
+will be configurable, and should be set to a value that will allow for
+corrective action to take place before cluster nodes lag or crash.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Device Health
+
+The Gluster operator will monitor the health of all devices in an automated
+Gluster cluster. It will report a "Warning" when a device reports failing health
+and will report a "Critical" problem when a device becomes inaccessible. Support
+for this feature may vary form one device type to another.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Volume Health
+
+The Gluster operator will monitor the health of all volumes in an automated
+Gluster cluster. It will report a "Warning" when a volume reports failing bricks
+and will report a "Critical" problem when a volume becomes inaccessible.
```

Comment:
It would be nice to monitor essential (glusterd, self-heal-daemon etc.) & optional services (quotad, bitrot-daemon) in addition to bricks as part of this activity.

Reply:
Besides failing bricks, it would be good to also check for unhealed files, or files in split-brain.